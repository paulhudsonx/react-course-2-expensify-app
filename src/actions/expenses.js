import database from "../firebase/firebase";

// Component calls action generator
// Action generator returns object
// Component dispatches object
// Redux store changes

// Component calls action generator
// Action generator returns object
// Component dispatches object
// Redux store changes

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    // return value is promise chain return value
    const result = database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
    console.log(result);
    return result;
  };
};

// Action generator function to removeExpense
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then((ref) => {
        dispatch(removeExpense({ id }));
      });
  };
};

export const updateExpense = ({ id } = {}, updates) => {
  console.log(`Updating ${id}`);
  return {
    type: "UPDATE_EXPENSE",
    id,
    updates,
  };
};

export const startUpdateExpense = ({ id } = {}, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then((ref) => {
        dispatch(updateExpense({ id }, updates));
      });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

/*
 * Fetches expense data from Firebase and sets them on the Redux store
 */
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    // return value is promise chain return value
    const result = database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then((snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
    console.log(result);
    return result;
  };
};
