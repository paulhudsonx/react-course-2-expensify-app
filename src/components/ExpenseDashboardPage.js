import React from 'react';
import ConnectedExpenseList from "./ExpenseList";
import ConnectedExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div>
    <ConnectedExpenseListFilters/>
    <ExpenseSummary />
    <ConnectedExpenseList />

  </div>
);

export default ExpenseDashboardPage;
