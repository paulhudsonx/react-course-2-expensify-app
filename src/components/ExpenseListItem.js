import React from "react";

import {Link} from 'react-router-dom';
import moment from "moment";
import {currencyFormat} from "../i18n/format";

export const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
    <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
    <p>
      {currencyFormat(amount / 100)}
      -
      {description}
      -
      {moment(createdAt).format(`MMMM Do, YYYY`)}</p>
  </div>
);

export default ExpenseListItem;
