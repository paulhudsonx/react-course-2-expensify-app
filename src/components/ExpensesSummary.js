import React from "react";

import {connect} from 'react-redux';

import expensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import {currencyFormat} from "../i18n/format";

export const ExpensesSummary = ({expenseCount, expensesTotal}) => (
  <div>
    {expenseCount === 0 && <h1>No matching expenses!</h1>}
    {expenseCount === 1 && <h1>Viewing {expenseCount} expense totalling {currencyFormat(expensesTotal / 100)}</h1>}
    {expenseCount > 1 && <h1>Viewing {expenseCount} expenses totalling {currencyFormat(expensesTotal / 100)}</h1>}
  </div>
)

const mapStateToProps = (state) => {

  const expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expensesTotal: expensesTotal(expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary)

