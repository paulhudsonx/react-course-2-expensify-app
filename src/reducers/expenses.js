const expensesReducerDefaultState = [];

/*
  Manipulates the expenses array
 */
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      console.log(`Removing action.id ${action.id}`);
      return state.filter(({ id }) => id !== action.id);
    case "UPDATE_EXPENSE":
      return state.map((e) => {
        if (e.id === action.id) {
          console.log(`Updating action.id ${action.id}`);
          console.log("Action updates +++");
          console.log(e);
          console.log(action);
          console.log({ ...e, ...action.updates });
          console.log("Action updates ---");
          return { ...e, ...action.updates };
        } else {
          return e;
        }
      });

    case "SET_EXPENSES":
      return action.expenses;

    default:
      return state;
  }
};

export default expensesReducer;
