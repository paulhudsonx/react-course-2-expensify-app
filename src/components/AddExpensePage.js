import React from 'react';

import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseFm";
import {addExpense} from "../actions/expenses";

export class AddExpensePage extends React.Component {

  onSubmit = (expense) => {
    console.log(expense)
    this.props.addExpense(expense)
    this.props.history.push("/")
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  // ... computed data from state and optionally ownProps
})

/*const mapDispatchToProps = (dispatch) => {
  return {
    onSubmit: (expense) => {
      dispatch(addExpense(expense))
    }
  }
}*/

export default connect(mapStateToProps, {addExpense})(AddExpensePage)

