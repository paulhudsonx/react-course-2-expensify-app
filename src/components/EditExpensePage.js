import React from 'react';

import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseFm";
import {removeExpense, updateExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
    console.log(expense);
    this.props.updateExpense({id: this.props.expense.id}, expense);
    this.props.history.push("/");
    console.log(`onSubmit called!`)
  }

  onRemove = (event) => {
    console.log(`Removing ${this.props.expense.id}`);
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push("/");
  }

  render() {
    return (
    <div>
      Editing the expense with id of {this.props.expense.id}
      <h1>Edit Expense</h1>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove}>Remove
      </button>
    </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  console.log(`Mapping ${props.match.params.id}`)
  return {
    expense: state.expenses.find((e) => {
      console.log(`Finding ${e.id}`);
      return e.id === props.match.params.id;
    })
  };
};

export default connect(mapStateToProps, {updateExpense, removeExpense})(EditExpensePage);
