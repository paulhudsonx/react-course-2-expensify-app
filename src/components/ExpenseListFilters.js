import React from "react";

import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filters";
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChanged = (calendarFocused) => {
    this.setState(() => ({calendarFocused}));
  };

  onTextChange = (event) => {
    console.log(event.target.value);
    this.props.setTextFilter(event.target.value);
  }

  onSortChange = (event => {
    const sortBy = event.target.value;
    switch (sortBy) {
      case 'date':
        this.props.sortByDate();
        break;
      case 'amount':
        this.props.sortByAmount();
        break;
      default:
        this.props.sortByDate();
        break;
    }
  })

  render() {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange}/>
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
          focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.onFocusChanged}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps, {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate})(ExpenseListFilters);
