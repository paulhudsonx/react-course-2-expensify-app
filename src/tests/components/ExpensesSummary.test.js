import React from "react";
import { shallow } from 'enzyme';

import {ExpensesSummary} from "../../components/ExpensesSummary";

import expenses from "../fixtures/expenses";

test(`should render ExpenseSummary correctly when 0 expenses`, () => {
  const param = {
    expenseCount: 0,
    expensesTotal: 0
  }
  const wrapper = shallow(<ExpensesSummary {...param}/>);
  expect(wrapper).toMatchSnapshot();
})

test(`should render ExpenseSummary correctly when 1 expenses`, () => {
  const param = {
    expenseCount: 1,
    expensesTotal: 25477
  }
  const wrapper = shallow(<ExpensesSummary {...param}/>);
  expect(wrapper).toMatchSnapshot();
})

test(`should render ExpenseSummary correctly when >1 expenses`, () => {
  const param = {
    expenseCount: 10,
    expensesTotal: 123123
  }
  const wrapper = shallow(<ExpensesSummary {...param}/>);
  expect(wrapper).toMatchSnapshot();
})
