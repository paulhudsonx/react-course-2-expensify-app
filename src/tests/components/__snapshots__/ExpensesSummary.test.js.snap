// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly when >1 expenses 1`] = `
<div>
  <h1>
    Viewing 
    10
     expenses totalling 
    $1,231.23
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly when 0 expenses 1`] = `
<div>
  <h1>
    No matching expenses!
  </h1>
</div>
`;

exports[`should render ExpenseSummary correctly when 1 expenses 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense totalling 
    $254.77
  </h1>
</div>
`;
